pkg/query-service/app/queryBuilder/functions.go

```diff
@@ -0,0 +1,286 @@
+					ewma = point.Value
+					initialized = true
+				}
+				// Continue until the EWMA is initialized
+				continue
+			}
+
+			if !math.IsNaN(point.Value) {
+				// Update EWMA with the current value
+				ewma = alpha*point.Value + (1-alpha)*ewma
+			}
+			// Set the EWMA value for the current point
+			series.Points[i].Value = ewma
+		}
+	}
+	return result
+}
+
+// funcMedian3 returns the median of 3 points for each point in a series
+func funcMedian3(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		median3 := make([]float64, 0)
+		for i := 1; i < len(series.Points)-1; i++ {
+			values := make([]float64, 0, 3)
+
+			// Add non-NaN values to the slice
+			for j := -1; j <= 1; j++ {
+				if !math.IsNaN(series.Points[i+j].Value) {
+					values = append(values, series.Points[i+j].Value)
+				}
+			}
+
+			// Handle the case where there are not enough values to calculate a median
+			if len(values) == 0 {
+				median3 = append(median3, math.NaN())
+				continue
+			}
+
+			median3 = append(median3, median(values))
+		}
+
+		// Set the median3 values for the series
+		for i := 1; i < len(series.Points)-1; i++ {
+			series.Points[i].Value = median3[i-1]
+		}
+	}
+	return result
+}
+
+// funcMedian5 returns the median of 5 points for each point in a series
+func funcMedian5(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		median5 := make([]float64, 0)
+		for i := 2; i < len(series.Points)-2; i++ {
+			values := make([]float64, 0, 5)
+
+			// Add non-NaN values to the slice
+			for j := -2; j <= 2; j++ {
+				if !math.IsNaN(series.Points[i+j].Value) {
+					values = append(values, series.Points[i+j].Value)
+				}
+			}
+
+			// Handle the case where there are not enough values to calculate a median
+			if len(values) == 0 {
+				median5 = append(median5, math.NaN())
+				continue
+			}
+
+			median5 = append(median5, median(values))
+		}
+
+		// Set the median5 values for the series
+		for i := 2; i < len(series.Points)-2; i++ {
+			series.Points[i].Value = median5[i-2]
+		}
+	}
+	return result
+}
+
+// funcMedian7 returns the median of 7 points for each point in a series
+func funcMedian7(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		median7 := make([]float64, 0)
+		for i := 3; i < len(series.Points)-3; i++ {
+			values := make([]float64, 0, 7)
+
+			// Add non-NaN values to the slice
+			for j := -3; j <= 3; j++ {
+				if !math.IsNaN(series.Points[i+j].Value) {
+					values = append(values, series.Points[i+j].Value)
+				}
+			}
+
+			// Handle the case where there are not enough values to calculate a median
+			if len(values) == 0 {
+				median7 = append(median7, math.NaN())
+				continue
+			}
+
+			median7 = append(median7, median(values))
+		}
+
+		// Set the median7 values for the series
+		for i := 3; i < len(series.Points)-3; i++ {
+			series.Points[i].Value = median7[i-3]
+		}
+	}
+	return result
+}
+
+func median(values []float64) float64 {
```

Comment:
this is true assuming `values` are sorted. Where is this asserted?

Reply:
No, there is a sort `sort.Float64s(values)` just below this line

Reply:
ah..my bad